Move edit/email endpoint to its own controller

monfresh opened the pull request:
**Why**: To keep concerns separated. Each updatable attribute has its
own rules and logic. Having a separate controller for each attribute
makes it easier to make changes in the future, and prevents having
conditionals that will always return the same value.

For example, when updating a user's email, we know that their phone
number cannot change, so we don't need to ask whether the phone
number has changed each time. This conditional should only be asked
when updating the phone number.

monfresh commented:
This is commit 1 of 2. The second commit will move the `edit/mobile` endpoint.
cc @brendansudol

sbc100 commented:
can you confirm that this change just moves code around?  No other changes?

if so, lgtm.

monfresh commented:
Yes, pretty much, although moving code around also means removing code that's no longer necessary. For example, by separating the concerns, we no longer need the `UserFlashUpdater` class anymore. Also, each controller now has its own logic, so the original logic wasn't just moved from one place to another, it was refactored to make sense for each controller.

monfresh commented:
It would be good either way to review my work, please, to make sure I didn't mess anything up while reorganizing the code.

monfresh commented:
Any other thoughts on this?

sbc100 commented:
looks good, reviewing details now.

sbc100 commented on app/controllers/users/edit_email_controller.rb:
```diff
@@ -0,0 +1,43 @@
+    end
+
+    def track_email_change
+      if !@update_user_email_form.email_taken?
```
Remove the `!` and flip the conditional blocks?

monfresh replied:
The reason for this particular order is that the most common path will be the first one, so the code will only have to check one conditional most of the time. If we flipped them, the first conditional would be false most of the time, and so we would be unnecessarily hitting 2 conditionals most of the time.

If the `!` bothers you that much, I can add a new method called `email_not_taken?`. Thoughts?

sbc100 replied:
I don't understand.  There is only one condition here.  Putting the most frequently hit block first it not (IIUI) an optimization that make sense in ruby, right?  Even in C, the best thing you might do is use a hint to the branch predictor, but that would only make sense in the lowest of lowest level code.

monfresh replied:
You're right. I just run a benchmark test, and there's no significant difference.

monfresh replied:
Although the version where the condition returns true in the first block is slightly faster.

sbc100 commented on app/controllers/users/edit_mobile_controller.rb:
```diff
@@ -0,0 +1,39 @@
+  class EditMobileController < ApplicationController
+    include PhoneConfirmation
+
+    before_action :confirm_two_factor_authenticated
```
In general I wonder if we should be using `prepend_before_action` here and elsewhere, but that would be a separate PR

monfresh replied:
Can you explain why? I don't think it matters, but maybe I'm missing something. `prepend_before_action` allows the method to run before any other `before_actions` that are defined. In most controllers where we use `confirm_two_factor_authenticated`, that's the only `before_action` we use, so there is no order to worry about.

sbc100 replied:
I just think we should be careful that the auth actions always happen first.. just in case one of the other actions happens to perform an action as the user (who might not be 2fa'd).  Unlikely I know.

sbc100 commented on app/controllers/users/edit_mobile_controller.rb:
```diff
@@ -0,0 +1,39 @@
+        flash[:notice] = t('devise.registrations.mobile_update_needs_confirmation')
+        prompt_to_confirm_mobile(@update_user_mobile_form.mobile)
+        return
+      end
```
Use if/else here? 

As a side note, I wonder if we should rename `prompt_to_confirm_mobile` -> `redirect_to_phone_confirmation`?

sbc100 commented on app/forms/update_user_email_form.rb:
```diff
@@ -28,16 +32,16 @@ def valid_form?
-  def mobile_changed?
-    false
+  def email_changed?
+    @email_changed == true
```
Should there just be attr_accessors?  (Separate PR maybe?)

monfresh replied:
Sure, I can do that in a separate PR.

sbc100 commented on config/routes.rb:
```diff
@@ -30,8 +30,10 @@
+    get '/edit/email' => 'users/edit_email#edit'
+    match '/edit/email' => 'users/edit_email#update', via: [:patch, :put]
+    get '/edit/mobile' => 'users/edit_mobile#edit'
+    match '/edit/mobile' => 'users/edit_mobile#update', via: [:patch, :put]
```
I think these should go in the "non-devise" routes below? IIUC devise_scope is for controllers that inherit from DeviseController, which these do not (see comment above).

monfresh replied:
Good catch!

sbc100 commented on app/controllers/users/edit_email_controller.rb:
```diff
@@ -0,0 +1,43 @@
+    end
+
+    def update
+      @update_user_email_form = UpdateUserEmailForm.new(current_user)
```
Now that we have a 1-to-1 controller -> form mapping should we just call this `@form`?

monfresh replied:
I prefer to be explicit, so that if I need to find a particular form in my text editor, I can find the specific form as opposed to getting multiple results for `@form`.

monfresh replied:
Also, when reading a particular view file, or jumping back and forth between various similar views, it might not be immediately clear which view you're looking at, so having a well-named variable goes a long way I think.

monfresh replied:
Excerpt from [Clean Code](https://www.amazon.com/Clean-Code-Handbook-Software-Craftsmanship-ebook/dp/B001GSTOAM#nav-subnav) by Robert "Uncle Bob" Martin:

> Use Intention-Revealing Names
It is easy to say that names should reveal intent. What we want to impress upon you is that we are serious about this. Choosing good names takes time but saves more than it takes. So take care with your names and change them when you find better ones. Everyone who reads your code (including you) will be happier if you do.
The name of a variable, function, or class, should answer all the big questions. It should tell you why it exists, what it does, and how it is used. If a name requires a comment, then the name does not reveal its intent.

sbc100 replied:
ok

monfresh commented:
@sbc100 I moved the routes to the non-Devise section, and updated the conditional in `EditMobileController#process_updates` to use if/else instead of return. Let me know if you want me to create that `email_not_taken?` method to get rid of the exclamation point.

sbc100 commented:
lgtm % our debate about branch prediction.

monfresh commented:
I updated the conditional. Should be good to go now.
